lib/librato/metrics/aggregator.rb

```diff
@@ -52,20 +52,33 @@ def initialize(opts={})
       # @return [Aggregator] returns self
       def add(measurements)
         measurements.each do |metric, data|
+          entry = {}
           if @prefix
             metric = "#{@prefix}.#{metric}"
           end
+          entry[:name] = metric.to_s
           if data.respond_to?(:each) # hash form
+            validate_parameters(data)
             value = data[:value]
             if data[:source]
-              metric = "#{metric}#{SOURCE_SEPARATOR}#{data[:source]}"
+              metric = "#{metric}#{SEPARATOR}#{data[:source]}"
+              entry[:source] = data[:source].to_s
+            end
+            if data[:tags] && data[:tags].respond_to?(:each)
+              metric = metric.to_s
+              data[:tags].sort.each do |key, value|
+                metric = "#{metric}#{SEPARATOR}#{key}#{SEPARATOR}#{value}"
+              end
+              entry[:tags] = data[:tags]
             end
           else
             value = data
           end
 
-          @aggregated[metric] ||= Aggregate.new
-          @aggregated[metric] << value
+          @aggregated[metric] = {} unless @aggregated[metric]
+          @aggregated[metric][:aggregate] ||= Aggregate.new
+          @aggregated[metric][:aggregate] << value
+          @aggregated[metric].merge!(entry)
```

Comment:
Gotcha yeah, I think I misunderstood how this was changing. Before this was a single value because source was serialized into the metric name to make a unique key but it looks like now you serialize the tags or source in and store it in the hash. Does it make sense to serialize them back out to reduce double-storage in the data structure?